### ISSUE TITLE
Virtual Energy Devices not updating in GUI anymore

### ISSUE THREAD

**Issue by akamming** [domoticz/domoticz#3617]
I noticed that in latest beta (I first observed this behaviour when i upgraded to beta 4.11290) my virtual energy devices (which i use for calculating real energy consumption using my solar inverter and P1 meter measurements) are not updated automatically in the GUI anymore.  

The virtual devices are updated by a dzvents script which is triggered everytime the P1 meter gives a reading using statements like

```
local SV=domoticz.devices("Stroom Verbruik")
SV.updateElectricity(vermogen,virtuelemeterstand) 
```

so everytime there is an update of the P1 meter (in my case every 10 seconds), half a second later there should be an update of the virtual device. What I now see is that the P1 meter is updated almost instantly in the new beta (nice!), but the virtual device is no longer updated until i hit F5 in the browser (tested on Chrome, FF and Edge. They all have the same behaviour).

**Comment by Reneke43** [comment]
I’ve got the same issue. Actually, none of the buttons are updating. After pushing for example a light button, the light switches on, but the GUI still says it’s off.

**Comment by gvamporov** [comment]
Same issue - virtual energy devices are not updating on web GUI. Tested on FF & Chrome with beta 4.11301. In addition my virtual temperature devices are not updating also. After refreshing the page, the devices are updated correctly. Deleting the cache did not solve this for me.
I guess this is a problem with virtual devices as a general...

**Comment by akamming** [comment]
@gvamporov i think you are right. I just checked my other virtual device (type general, percentage) and it indeed has the same behaviour. Just to be sure i just upgraded to 11.301 and still have the issue

**Comment by gizmocuz** [comment]
We know where the problem is. It is in the way Lua/dzVents are updating internal sensors, there are directly poked into the database and do not follow common internal routes.

**Comment by gizmocuz** [comment]
@gvamporov, If you update the sensors via the JSON API, there should be no issue, so no it is not a general virtual sensor issue

**Comment by gizmocuz** [comment]
Is it possible to test beta version 11305

**Comment by gvamporov** [comment]
> @gvamporov, If you update the sensors via the JSON API, there should be no issue, so no it is not a general virtual sensor issue

I'm updating them with Lua/dzVents script, so i guess you are correct about the issue.
Just updated to 11305: the temperature sensors are updating now, but not the electric (instant+counter if i remember correctly).
Edit: percentage virtual device is not updating either...

**Comment by gizmocuz** [comment]
@gvamporov , I tested only with a temperature sensor... I think it needs more work

**Comment by gizmocuz** [comment]
@gvamporov , just tested with a virtual percentage sensor and it works normally, are you sure ???

script:

```
return {
	on = {
		timer = {'every minute'},
    },
	execute = function(domoticz, timer)
	    local myTemp=domoticz.devices("MyPercentage")
	    myTemp.updatePercentage(12.3)
		domoticz.log('Timer event was triggered by ' .. timer.trigger, domoticz.LOG_INFO)
	end
}

```

**Comment by gvamporov** [comment]
> @gvamporov , just tested with a virtual percentage sensor and it works normally, are you sure ???


Yes, web GUI is not updating percentage devices on my setup...
I just checked my scripts and my percentage/electric devices are updated with Lua scripts, and my temperature devices are updated with dzVents script. 
But i am not sure if this is related to the problem...

**Comment by akamming** [comment]
just installed the new beta (4.11307). Great work. Energy devices are now working fine ;)

Percentage i will check tomorrow. Solar energy production just stopped for today, so that script does not calculate anymore, i have to await the sun to be shining again :-)

**Comment by gizmocuz** [comment]
Thanks for the feedback.. building took a long time 😁👍 hope all is ok now

**Comment by Reneke43** [comment]
I've just installed beta 4.11307, and it's indeed working again. Thanx a lot. 👍🏻

**Comment by Reneke43** [comment]
Hmmm, beta 4.11307 introduced a new issue. My blockly scripts are no longer working. For example. When a hue motion sensor detects motion, a blockly script switches on a coco enabled toilet fan.
Direct hue switching or coco switching from Domoticz works as expected. So some goes wrong at script level.

Edit: I just found out the response is extremely slow, i.e. 2 minutes between action and reaction.

**Comment by gvamporov** [comment]
I updated to 11307 and still percentage & electric are not updating for me (both updated via Lua script).

**Comment by gizmocuz** [comment]
@Reneke43 , i think you have a kind of internal 'loop'

@gvamporov , please post a simpel lua script that should update a percentage sensor in Lua every minute, that way i should be able to replicate it

@Reneke43 , please post a screenshot of a minimal blockly script that listens to a "ON" command from a dummy, and switches ON another dummy (this should not work in your case...)

**Comment by akamming** [comment]
My inverter just went out of night mode, so the other virtual devices now also started getting new values from my Energy dzvents script. 

I am happy to report that all my virtual devices (Energy, Percentage, Temperature) are all updating correctly (continously and instantly!) in the GUI. 

@gizmocuz thank for the good work!

**Comment by gvamporov** [comment]
@gizmocuz , here are 2 examples (Lua & dzVents):

```
commandArray = {}
 if ( devicechanged['Battery Amps'] or devicechanged['PV Amps']) then
     commandArray['UpdateDevice']="68|0|50"
 end
 return commandArray
```

```return {
	on = {
		devices = { 'Battery Amps', 'PV Amps' }
	},
	data = {},
	logger = {},
	execute = function(domoticz, triggeredItem)
	    domoticz.devices(68).updatePercentage(50)
	end
}
```

Lua script is not updating web GUI, but dzVents works as expected...

**Comment by Reneke43** [comment]
@gizmocuz 
<img width="478" alt="Schermafbeelding 2019-09-12 om 09 33 00" src="https://user-images.githubusercontent.com/19175257/64763538-80616300-d540-11e9-85ec-ad9ac6897a3b.png">

In the above image "Keuken lichtschakelaar" is a click-on, click-off switch. As I can see on the switches page, this switches state changes are registered in real time.
"Keukenlicht dimmer" is a group of two HUE lamps (defined on the HUE bridge). This devices is also realtime switchable from domoticz. 

But the above script no longer works, while it has for the last 2 years ;-) 
Actually, it has been working up until 4.11305

**Comment by gizmocuz** [comment]
@Reneke43 , going to debug this... 

First test (more will follow later today), simple 2 dummy switches, with this blockly:

![image](https://user-images.githubusercontent.com/2265299/64769741-223a7d00-d54c-11e9-8122-46ede98944ee.png)

Works without any issue

**Comment by gizmocuz** [comment]
Created a "Group" with "Dummy_Switch_2" in it, switching the Group on/off also switches the Dummy_Switch_2 On/Off, next is in Blockly

**Comment by gizmocuz** [comment]
Okey, so this also works:

![image](https://user-images.githubusercontent.com/2265299/64769929-7c3b4280-d54c-11e9-9ce3-0f5bff55e6aa.png)

Your blockly looks wrong ! If "Keukentafel dimmer" is a group, then you should delete it from your blockly, from the blockly devices, select the Scenes/Groups and add the group in the correct way

(Do note the different type/color)

Else please PM me on the Forum so I could send you a debug request for your database backup, do not attach it here)

**Comment by Reneke43** [comment]
I've tried the dummy setup, but it's not working. When switching on Dummy1, Dummy2 stays off.

However, in the logs I found a message that the LUA script I'm using to calculate my real energy usage (Grid and PV combined) is the trouble maker (Not responding for more than 10 seconds) After disabling this script, everything else works fine. So I've got to do some trouble shooting in that script.

Thanx very much for the effort.

By the way, my blocky isn't wrong. "Keukentafel dimmer" is a group on my HUE bridge, which is presented as a single device in Domoticz.

**Comment by gizmocuz** [comment]
@Reneke43 , thanks for clearing this up, hope you can find the issue with your lua script...
If not, please post it here if possible

Your correct about the Hue group switch, confusion all over ;)

**Comment by rwaaren** [comment]
"there have been some recent changes,not to values, but to the UpdateDevice call (now using the MainWorker::UpdateDevice call)"

@gizmocuz , does my initial guess on this make sense ?

from [forum link](https://www.domoticz.com/forum/viewtopic.php?f=6&t=29371&p=223897#p223897l)

"Have to do some more investigation and testing but my initial guess is that the way incremental counters are updated is now conforming to what the devicetype name implies. This results in that the updateCounter method in dzVents does no longer simply replace the value but is added to the current value like the incrementCounter method already did."

**Comment by GravityRZ** [comment]
i think in the old way it was used as a counter(not incremental) and the incremental part was displayed as usage of that day in kWh

i made a normal counter but that one display total counter amounts 

what counter do  i need to use which displays counter amount in kWh an kWH usage of that day on top, just like the old counter+incremenal did?

![image](https://user-images.githubusercontent.com/16853535/64844345-ccc3a600-d607-11e9-9a31-726710dace5f.png)

**Comment by rwaaren** [comment]
> i think in the old way it was used as a counter(not incremental) 

I think the same but I am looking @gizmocuz to get this confirmed.
Best to use the forum for the question on what to use now.  On the forum many more users will benefit from the discussion and the result (if any)

**Comment by GravityRZ** [comment]
yes but since this was related to an update made here i needed to know if this was a bug or intentional. if this is intentional then i will post on the forum to ask for som e best practices. maybee not a lot of people are experiencing this a problem yet.

**Comment by gizmocuz** [comment]
@rwaaren , the way the counter incremental was (original) created, is that you could use it like this
Original value 0

- Send 10, current value 10
- Send 3, current value 13
- Send 2, current value 15

That is why it is called a incremental counter

A normal counter, just updates the value (replaces the value)

**Comment by rwaaren** [comment]
> That is why it is called a incremental counter
> A normal counter, just updates the value (replaces the value)

Thx for the confirmation.
Because the updateDevice function from eventSystem.cpp did not increment but replaced, some scripts that did use that incorrect mechanism now produce unexpected values and need to change. 
I am working on that now.

**Comment by GravityRZ** [comment]
and if we look at it another way.
we obviously need a meter which also counts the usage of that day, every day.
could we not use that functionality of the old "counter+ incremental" and make a new virtual device "counter+daily usage)

the incremental part is not the problem because we can get rid of that by using the normal counter.

**Comment by gizmocuz** [comment]
@GravityRZ , we have a normal meter for this already

**Comment by GravityRZ** [comment]
which meter?

**Comment by gizmocuz** [comment]
Have you tried a normal meter ? Instead of a counter incremental type ?

**Comment by GravityRZ** [comment]
Yes i have but then i loose the daily usage in kWh at the top of the meter

**Comment by gvamporov** [comment]
I'm with 11316 now and still see the same with virtual devices that are updated with LUA script - they don't get updated on web GUI.
I had time and migrated one of the scripts from LUA to dzVents and now the relevant device is updated on the GUI as expected...

**Comment by gizmocuz** [comment]
@gvamporov , this have been tested and working. You should start with a very simple dummy hardware device, a very simple lua script that updates it.
(simple like , if switch is pressed, set percentage to a value, or a time based script that does this)

**Comment by gizmocuz** [comment]
@gvamporov , for example the below script updates a percentage sensor (with index 1414) based on a switch state

```
commandArray = {}

for deviceName,deviceValue in pairs(devicechanged) do
    if (deviceName=='MySwitch') then
        if deviceValue == "On" then
            print("Device is On")
            commandArray[0]={['OpenURL']="http://127.0.0.1:8080/json.htm?type=command&param=udevice&idx=1414&nvalue=0&svalue=43.5" }
        elseif deviceValue == "Off" then
            print("Device is Off")
            commandArray[0]={['OpenURL']="http://127.0.0.1:8080/json.htm?type=command&param=udevice&idx=1414&nvalue=0&svalue=23.4" }
        end
    end
end

return commandArray

```

Maybe this could have been done simpler, but it's just for demonstration purpose

**Comment by gvamporov** [comment]
@gizmocuz i did send you example of very simple script, that does not work.
The only difference from your example, as far as i can tell, is the way virtual devices are updated: 
commandArray['UpdateDevice'] vs. commandArray[0]={['OpenURL']
This was a problem with dzVents scripts, but they got fixed and LUA script are not yet fixed.
Again, my scripts are working - if i refresh the whole page, i see that the virtual devices are updated...

**Comment by GravityRZ** [comment]
> @GravityRZ , you can also use a kWh meter type ? You should probably have a few already when using zwave
> 
> @gvamporov , this have been tested and working. You should start with a very simple dummy hardware device, a very simple lua script that updates it.
> (simple like , if switch is pressed, set percentage to a value, or a time based script that does this)

@gizmocuz 
i removed the incremental counters and used standard counters.
the counters work and now they seem to have daily usage on top.

i am pretty sure i tested it several times before and they did not have that.
before they just cpied the total amount.

has anything changed lately

**Comment by gizmocuz** [comment]
@GravityRZ , you could have 'simply' changed the type of incremental to the normal one instead of deleting it

**Comment by GravityRZ** [comment]
> @GravityRZ , you could have 'simply' changed the type of incremental to the normal one instead of deleting it

not sure if i follow you.
there is no option in counter incremental or counter to switch the type, only the type of data (energy,gas,water,counter,energy generated,time)

**Comment by gizmocuz** [comment]
In the gui, you should be able to 'edit' the counter device and change it's type (you can also make it a water meter for instance)

**Comment by GravityRZ** [comment]
yes but i can not change a counter incremental into  a counter and that was exactly the problem.

**Comment by gizmocuz** [comment]
@GravityRZ , as mentioned before, you could have, maybe you have a backup ? See below image:

![image](https://user-images.githubusercontent.com/2265299/65157508-20335b00-da31-11e9-9e93-460c0174818c.png)

**Comment by GravityRZ** [comment]
i know but you are missing the point

You can not turn an existing Counter+incremental(test) into a counter(test2)
sure you can change the type, i know 


![image](https://user-images.githubusercontent.com/16853535/65163193-d64f7280-da3a-11e9-844b-d0caa59bf3ae.png)

**Comment by gizmocuz** [comment]
@GravityRZ , i don't think i miss any point ... but there seems to be two ways to add a counter.. you can add a counter and change it's type to energy/counter/gas/water.. and counter incremental.
Seems very confusing...
But this is not the topic to discuss a counter incremental here...
I am going to cleanup this thread a bit, I will try and fix the changing of the meter type

**Comment by gizmocuz** [comment]
@gvamporov , found the issue! (beta 11324)

Using Lua with commandArray['UpdateDevice'] will now also call all pushers (including updating the gui via websockets)

Please let me know if this issue can be closed

**Comment by svde** [comment]
I'm not sure if its related, but all icons aren't being updated with 11324. This has been going on for +/- 1 week. Hue, zwave, dummy switches, etc are all affected.

**Comment by gizmocuz** [comment]
@svde, sorry, what do you mean with 'icons' not updating ? Could you provide a step-by-step instruction so this can be replicated ?

**Comment by svde** [comment]
> 
> 
> @svde, sorry, what do you mean with 'icons' not updating ? Could you provide a step-by-step instruction so this can be replicated ?

Click the icon of this [device](https://www.eijk.nu/se/Annotation%202019-09-20%20102649.png), it says "switching on", but then the icon remains the same. Only after refreshing (F5) does the icon change to reflect the current state ([on](https://www.eijk.nu/se/Annotation%202019-09-20%20103157.png)). Note that this is a Zwave dimmer.

I think I found it... I'm accessing it through an apache reverse proxy, accessing it directly does work. My hunch is that something recently changed which needs adjusting in the reverse proxy config.

**Comment by gvamporov** [comment]
@gizmocuz just updated to 11331 and now it looks Ok. LUA script are updating web GUI... Thanks! :)

@svde it seems to be related to recent migration to websockets. Here is what worked for me for apache proxy:
```
        RewriteEngine On
        RewriteCond %{QUERY_STRING} transport=polling       [NC]
        RewriteRule /ha/(.*) http://192.168.123.10:8080/$1 [P]
        RewriteCond %{HTTP:Upgrade} websocket               [NC]
        RewriteRule /ha/(.*) ws://192.168.123.10:8080/$1 [P]
```
From outside, I am accessing it via https://my.domain/ha/

There is a [forum thread](https://www.domoticz.com/forum/viewtopic.php?f=6&t=29291&sid=fab0b5e2e6a5ae4a13cef2b001d5bc83) from nginx user, that helped me find the solution...

**Comment by gizmocuz** [comment]
Previous issue is solved/handled in beta 11332 and up
Closing this issue
